### PR TITLE
New campaign action:  Send a webhook

### PR DESCRIPTION
**Pull request by kuzmany**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/204
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/854
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Just add new action with curl call to url
Return true If response code = 200 or 201.

Note: this PR contains #4355 which should be merged first.

Thanks to @Loanatik for donate this feature.

image

#### Steps to test this PR:
1. Basic test could be create test.php anywhere and call GET or POST there.
2. Extended test could be with service external service.  (I tested with our private address validator service and works with POST request data and headers). Maybe people from https://github.com/mautic/mautic/issues/854 could test

### CODE REVIEW

**Comment by Loanatik** [comment]
This is brilliant!

File: app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php [hunk 1]
```diff
@@ -50,13 +60,71 @@ public function __construct(IpLookupHelper $ipLookupHelper, AuditLogModel $audit
...
+                    $headers,
+                    $timeout,
+                    $data
+                );
+            }
+            if (in_array($response->code, [200, 201])) {
+                return $event->setResult(true);
+            }
+        } catch (\Exception $e) {
+        }
```
Comment (escopecz) [inline app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php]: We have to let the users know why the request failed. It fails silently without any clue of what happened. Whether it returned 500 or timeouted. Could you please add some error messages so the users won't report errors like "the campaign URL call doesn't work", but give them information so they could debug themselves?

You can inspire here: https://github.com/mautic/mautic/blob/staging/app/bundles/PluginBundle/EventListener/CampaignSubscriber.php#L116

File: app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php [hunk 2]
```diff
@@ -33,15 +37,21 @@ class CampaignSubscriber extends CommonSubscriber
...
     /**
+     * @var Http
+     */
+    protected $connector;
+
+    /**
      * CampaignSubscriber constructor.
      *
      * @param IpLookupHelper $ipLookupHelper
      * @param AuditLogModel  $auditLogModel
```
Comment (escopecz) [inline app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php]: Add the new dependency to the annotation please.

File: app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php [hunk 3]
```diff
@@ -50,13 +60,71 @@ public function __construct(IpLookupHelper $ipLookupHelper, AuditLogModel $audit
...
+            $headers[trim($key)] = trim($value);
+        }
+
+        try {
+            $method = $config['method'];
+            $data   = !empty($config['additional_data']['list']) ? $config['additional_data']['list'] : '';
+            $data   = array_flip(AbstractFormFieldHelper::parseList($data));
+            if (in_array($method, ['get', 'trace'])) {
+                $response = $this->connector->$method(
+                    $config['url'],
```
Comment (escopecz) [inline app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php]: Maybe the `$data` should be added as a URL query params in this case?

**Review by escopecz — CHANGES_REQUESTED** [review]
Thanks Zdeno! This feature has been requested many times.

I added some thoughts to the new code, but over all, it looks good to me! Please review the comments and implement them if they make sense to you.

File: app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php (hunk 2 above)
Comment (kuzmany) [inline app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php]: Fixed

File: app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php [hunk 4]
```diff
@@ -50,13 +60,73 @@ public function __construct(IpLookupHelper $ipLookupHelper, AuditLogModel $audit
...
+                $response = $this->connector->$method(
+                    $url,
+                    $headers,
+                    $timeout,
+                    $data
+                );
+            }
+            if (in_array($response->code, [200, 201])) {
+                return $event->setResult(true);
+            }
```
Comment (escopecz) [inline app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php]: I'd also add something like
```
else {
    $event->setFailed('Response returned HTTP code '.$response->code);
}
```
I'm not sure whether the error is run through the translator later. Maybe you'll have to add Translator as a dependency to this Subscriber.

File: app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php [hunk 5]
```diff
@@ -50,13 +60,73 @@ public function __construct(IpLookupHelper $ipLookupHelper, AuditLogModel $audit
...
+        } catch (\Exception $e) {
+            return $event->setResult($e->getMessage());
```
Comment (escopecz) [inline app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php]: The method should be `setFailed` instead of `setResult`.

**Review by escopecz — COMMENTED** [review]
It was fast!

File: app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php (hunk 1 above)
Comment (kuzmany) [inline app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php]: Improved. Please check it

**Comment by Loanatik** [comment]
Should have user friendly description:
image

File: app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php (hunk 4 above)
Comment (escopecz) [inline app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php]: Ok, should be resolved with the previous commit

File: app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php [hunk 6]
```diff
@@ -50,13 +60,73 @@ public function __construct(IpLookupHelper $ipLookupHelper, AuditLogModel $audit
...
+     *
+     * @return CampaignExecutionEvent
+     */
+    public function onCampaignTriggerAction(CampaignExecutionEvent $event)
+    {
+        if (!$event->checkContext('campaign.remoteurl')) {
+            return;
+        }
+        $config  = $event->getConfig();
+        $timeout = 10;
```
Comment (Loanatik) [inline app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php]: How does timeout get set to the value specified by the user?

Comment (kuzmany) [inline app/bundles/CampaignBundle/EventListener/CampaignSubscriber.php]: Yes, already update with timeout option. Check screenshot

image

**Comment by Loanatik** [comment]
Rather than just an authorization header, wouldn't make more sense to allow any custom header similar to the way you can add any data?

**Comment by kuzmany** [comment]
@Loanatik please send examples, I don't know about it

**Comment by Loanatik** [comment]
I don't personally have a big need for this, though I'd imagine others will:
image

I have not tested the data section for tokens.  Am I correct in assuming that that they should work?

**Comment by kuzmany** [comment]
@Loanatik contacts fields tokens coming soon I will add it
Headers: I think It works now but I need look to documentation

**Comment by Loanatik** [comment]
Hi Z-

    There are a whole slew of improvements that I would like to see in Mautic once #4357 is complete.   Do you want to work on another project?  If so, I think merging duplicates needs improvement .  Is that an area you consider important?

Corey


[LOANATIK_ID_Web]
8520 East Shea Blvd #115, Scottsdale, AZ 85260
www.loanatik.com<http://www.loanatik.com/>

Corey Schwartz
480-588-0397<tel:480-588-0397>
602-739-7060<tel:602-739-7060> (m)
480-696-6495 (f)
cschwartz@Loanatik.com<mailto:cschwartz@Loanatik.com>

Loanatik, l.l.c., 8520 E Shea Blvd #115, Scottsdale, AZ 85260.  www.loanatik.com. 480-568-0333. Copyright 2015. All Rights Reserved. This is not an offer to enter into a loan or agreement. Not all borrowers will qualify. Information, rates and programs are subject to change without prior notice. All products are subject to credit and property approval. Only available in CO & TX. Other restrictions and limitations apply. Corporate NMLS ID #1328447 | NMLS Consumer Access<http://www.nmlsconsumeraccess.org/>
State Licenses: Colorado & Texas License Issued.

**Comment by kuzmany** [comment]
Added  headers and support for contactfield tokens for data and headers
image

**Comment by kuzmany** [comment]
Added documentation.
Now ready for tests.

**Comment by escopecz** [comment]
Note: this PR contains https://github.com/mautic/mautic/pull/4355 which should be merged first.

**Review by alanhartless — CHANGES_REQUESTED** [review]
Good stuff! I think instead of naming this "Call remote url" let's call it "Send a webhook"

**Comment by Loanatik** [comment]
I've been testing this PR and it works as described.

**Comment by calevans** [comment]
Do we have a status on this? Any idea when it will be included?

Cheers! :)
=C=

**Comment by kuzmany** [comment]
@calevans Mautic 2.10

**Comment by kuzmany** [comment]
Please re-test. 
Refactor to Send a webhook and moved to MauticWebhookBundle.

**Comment by escopecz** [comment]
It's failing some CI tests. Please look at those when you get a minute.

**Comment by kuzmany** [comment]
@escopecz morning. fixed

**Review by alanhartless — APPROVED** [review]
I removed trace and options as they are not applicable to webhooks and added the missing DELETE method to the method choices. I also added a not blank validator to ensure URL is required. Otherwise, good stuff!

**Review by escopecz — APPROVED** [review]
Works great. Amazing feature. Thanks! 👍

**Comment by ghost** [comment]
Does it support the payload like in the attached picture
Screenshot from 2020-05-21 21-20-37
Also how do i know the status / response of a api call in send a webhook

**Comment by dichvuhuuich** [comment]
> Does it support the payload like in the attached picture
> Screenshot from 2020-05-21 21-20-37
> Also how do i know the status / response of a api call in send a webhook

I want to know the status / response of a api call in send a webhook, too, do you know that?

**Pull request merged**